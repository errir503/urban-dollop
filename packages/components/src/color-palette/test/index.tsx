/**
 * External dependencies
 */
import { render, screen, within, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

/**
 * Internal dependencies
 */
import ColorPalette from '..';

const EXAMPLE_COLORS = [
	{ name: 'red', color: '#f00' },
	{ name: 'green', color: '#0f0' },
	{ name: 'blue', color: '#00f' },
];
const INITIAL_COLOR = EXAMPLE_COLORS[ 0 ].color;

function getWrappingPopoverElement( element: HTMLElement ) {
	return element.closest( '.components-popover' );
}

describe( 'ColorPalette', () => {
	it( 'should render a dynamic toolbar of colors', () => {
		const onChange = jest.fn();

		const { container } = render(
			<ColorPalette
				colors={ EXAMPLE_COLORS }
				value={ INITIAL_COLOR }
				onChange={ onChange }
			/>
		);

		expect( container ).toMatchSnapshot();
	} );

	it( 'should render three color button options', () => {
		const onChange = jest.fn();

		render(
			<ColorPalette
				colors={ EXAMPLE_COLORS }
				value={ INITIAL_COLOR }
				onChange={ onChange }
			/>
		);

		expect(
			screen.getAllByRole( 'button', { name: /^Color:/ } )
		).toHaveLength( 3 );
	} );

	it( 'should call onClick on an active button with undefined', async () => {
		const user = userEvent.setup();
		const onChange = jest.fn();

		render(
			<ColorPalette
				colors={ EXAMPLE_COLORS }
				value={ INITIAL_COLOR }
				onChange={ onChange }
			/>
		);

		await user.click(
			screen.getByRole( 'button', { name: /^Color:/, pressed: true } )
		);

		expect( onChange ).toHaveBeenCalledTimes( 1 );
		expect( onChange ).toHaveBeenCalledWith( undefined );
	} );

	it( 'should call onClick on an inactive button', async () => {
		const user = userEvent.setup();
		const onChange = jest.fn();

		render(
			<ColorPalette
				colors={ EXAMPLE_COLORS }
				value={ INITIAL_COLOR }
				onChange={ onChange }
			/>
		);

		// Click the first unpressed button
		// (i.e. a button representing a color that is not the current color)
		await user.click(
			screen.getAllByRole( 'button', {
				name: /^Color:/,
				pressed: false,
			} )[ 0 ]
		);

		// Expect the green color to have been selected
		expect( onChange ).toHaveBeenCalledTimes( 1 );
		expect( onChange ).toHaveBeenCalledWith( EXAMPLE_COLORS[ 1 ].color, 1 );
	} );

	it( 'should call onClick with undefined, when the clearButton onClick is triggered', async () => {
		const user = userEvent.setup();
		const onChange = jest.fn();

		render(
			<ColorPalette
				colors={ EXAMPLE_COLORS }
				value={ INITIAL_COLOR }
				onChange={ onChange }
			/>
		);

		await user.click( screen.getByRole( 'button', { name: 'Clear' } ) );

		expect( onChange ).toHaveBeenCalledTimes( 1 );
		expect( onChange ).toHaveBeenCalledWith( undefined );
	} );

	it( 'should allow disabling custom color picker', () => {
		const onChange = jest.fn();

		const { container } = render(
			<ColorPalette
				colors={ EXAMPLE_COLORS }
				disableCustomColors
				value={ INITIAL_COLOR }
				onChange={ onChange }
			/>
		);

		expect( container ).toMatchSnapshot();
	} );

	it( 'should render dropdown and its content', async () => {
		const user = userEvent.setup();
		const onChange = jest.fn();

		render(
			<ColorPalette
				colors={ EXAMPLE_COLORS }
				value={ INITIAL_COLOR }
				onChange={ onChange }
			/>
		);

		await user.click(
			screen.getByRole( 'button', {
				name: /^Custom color picker/,
				expanded: false,
			} )
		);

		const dropdownButton = screen.getByRole( 'button', {
			name: /^Custom color picker/,
			expanded: true,
		} );

		expect( dropdownButton ).toBeVisible();

		expect(
			within( dropdownButton ).getByText( EXAMPLE_COLORS[ 0 ].name )
		).toBeVisible();

		expect(
			within( dropdownButton ).getByText(
				EXAMPLE_COLORS[ 0 ].color.replace( '#', '' )
			)
		).toBeVisible();

		// Check that the popover with custom color input has appeared.
		const dropdownColorInput = screen.getByLabelText( 'Hex color' );

		await waitFor( () =>
			expect(
				getWrappingPopoverElement( dropdownColorInput )
			).toBePositionedPopover()
		);
	} );

	it( 'should show the clear button by default', () => {
		const onChange = jest.fn();

		render(
			<ColorPalette
				colors={ EXAMPLE_COLORS }
				value={ INITIAL_COLOR }
				onChange={ onChange }
			/>
		);

		expect(
			screen.getByRole( 'button', { name: 'Clear' } )
		).toBeInTheDocument();
	} );

	it( 'should show the clear button even when `colors` is an empty array', () => {
		const onChange = jest.fn();

		render( <ColorPalette colors={ [] } onChange={ onChange } /> );

		expect(
			screen.getByRole( 'button', { name: 'Clear' } )
		).toBeInTheDocument();
	} );
} );
