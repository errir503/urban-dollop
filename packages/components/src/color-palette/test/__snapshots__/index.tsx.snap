// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorPalette should allow disabling custom color picker 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: calc(4px * 3);
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-0>* {
  min-height: 0;
}

<div>
  <div
    class="components-flex components-h-stack components-v-stack emotion-0 emotion-1"
    data-wp-c16t="true"
    data-wp-component="VStack"
  >
    <div
      class="components-circular-option-picker"
    >
      <div
        class="components-circular-option-picker__swatches"
      >
        <div
          class="components-circular-option-picker__option-wrapper"
        >
          <button
            aria-label="Color: red"
            aria-pressed="true"
            class="components-button components-circular-option-picker__option is-pressed"
            style="background-color: rgb(255, 0, 0); color: rgb(255, 0, 0);"
            type="button"
          />
          <svg
            aria-hidden="true"
            fill="#000"
            focusable="false"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M16.7 7.1l-6.3 8.5-3.3-2.5-.9 1.2 4.5 3.4L17.9 8z"
            />
          </svg>
        </div>
        <div
          class="components-circular-option-picker__option-wrapper"
        >
          <button
            aria-label="Color: green"
            aria-pressed="false"
            class="components-button components-circular-option-picker__option"
            style="background-color: rgb(0, 255, 0); color: rgb(0, 255, 0);"
            type="button"
          />
        </div>
        <div
          class="components-circular-option-picker__option-wrapper"
        >
          <button
            aria-label="Color: blue"
            aria-pressed="false"
            class="components-button components-circular-option-picker__option"
            style="background-color: rgb(0, 0, 255); color: rgb(0, 0, 255);"
            type="button"
          />
        </div>
      </div>
      <div
        class="components-circular-option-picker__custom-clear-wrapper"
      >
        <button
          class="components-button components-circular-option-picker__clear is-tertiary"
          type="button"
        >
          Clear
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`ColorPalette should render a dynamic toolbar of colors 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: calc(4px * 3);
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-0>* {
  min-height: 0;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 0;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-2>* {
  min-height: 0;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: calc(4px * 0.5);
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-4>* {
  min-height: 0;
}

.emotion-6 {
  display: block;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

<div>
  <div
    class="components-flex components-h-stack components-v-stack emotion-0 emotion-1"
    data-wp-c16t="true"
    data-wp-component="VStack"
  >
    <div
      class="components-dropdown"
      tabindex="-1"
    >
      <div
        class="components-flex components-h-stack components-v-stack components-color-palette__custom-color-wrapper emotion-2 emotion-1"
        data-wp-c16t="true"
        data-wp-component="VStack"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          aria-label="Custom color picker. The currently selected color is called "red" and has a value of "#f00"."
          class="components-color-palette__custom-color-button"
          style="background: rgb(255, 0, 0);"
        />
        <div
          class="components-flex components-h-stack components-v-stack components-color-palette__custom-color-text-wrapper emotion-4 emotion-1"
          data-wp-c16t="true"
          data-wp-component="VStack"
        >
          <span
            class="components-truncate components-color-palette__custom-color-name emotion-6 emotion-1"
            data-wp-c16t="true"
            data-wp-component="Truncate"
          >
            red
          </span>
          <span
            class="components-truncate components-color-palette__custom-color-value components-color-palette__custom-color-value--is-hex emotion-6 emotion-1"
            data-wp-c16t="true"
            data-wp-component="Truncate"
          >
            #f00
          </span>
        </div>
      </div>
    </div>
    <div
      class="components-circular-option-picker"
    >
      <div
        class="components-circular-option-picker__swatches"
      >
        <div
          class="components-circular-option-picker__option-wrapper"
        >
          <button
            aria-label="Color: red"
            aria-pressed="true"
            class="components-button components-circular-option-picker__option is-pressed"
            style="background-color: rgb(255, 0, 0); color: rgb(255, 0, 0);"
            type="button"
          />
          <svg
            aria-hidden="true"
            fill="#000"
            focusable="false"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M16.7 7.1l-6.3 8.5-3.3-2.5-.9 1.2 4.5 3.4L17.9 8z"
            />
          </svg>
        </div>
        <div
          class="components-circular-option-picker__option-wrapper"
        >
          <button
            aria-label="Color: green"
            aria-pressed="false"
            class="components-button components-circular-option-picker__option"
            style="background-color: rgb(0, 255, 0); color: rgb(0, 255, 0);"
            type="button"
          />
        </div>
        <div
          class="components-circular-option-picker__option-wrapper"
        >
          <button
            aria-label="Color: blue"
            aria-pressed="false"
            class="components-button components-circular-option-picker__option"
            style="background-color: rgb(0, 0, 255); color: rgb(0, 0, 255);"
            type="button"
          />
        </div>
      </div>
      <div
        class="components-circular-option-picker__custom-clear-wrapper"
      >
        <button
          class="components-button components-circular-option-picker__clear is-tertiary"
          type="button"
        >
          Clear
        </button>
      </div>
    </div>
  </div>
</div>
`;
